packages/flutter/lib/src/services/platform_messages.dart

```diff
@@ -237,3 +250,647 @@ class PlatformMessages {
+              }
+          }
+        );
+        try {
+          await invokeMethod('listen', arguments);
+        } catch (e) {
+          PlatformMessages.setBinaryMessageHandler(name, null);
+          controller.addError(e);
+        }
+      }, onCancel: () async {
+        PlatformMessages.setBinaryMessageHandler(name, null);
+        try {
+          await invokeMethod('cancel', arguments);
+        } catch (exception, stack) {
+          FlutterError.reportError(new FlutterErrorDetails(
+            exception: exception,
+            stack: stack,
+            library: 'services library',
+            context: 'while de-activating platform stream on channel $name',
+          ));
+        }
+      }
+    );
+    return controller.stream;
+  }
+}
+
+/// A message encoding/decoding mechanism with support for method calls and
+/// enveloped replies.
+///
+/// Reply envelopes are binary messages with enough structure that the codec can
+/// distinguish between a successful reply and an error. In the former case,
+/// the codec must be able to extract the result, possibly `null`. In
+/// the latter case, the codec must be able to extract an error code string,
+/// a (human-readable) error message string, and a value providing any
+/// additional error details, possibly `null`. These data items are used to
+/// populate a [PlatformException].
+///
+/// All operations throw [FormatException], if conversion fails.
+abstract class MessageCodec {
+  /// [MessageCodec] with UTF-8 encoded JSON messages.
+  ///
+  /// Supported messages are acyclic values of these forms:
+  ///
+  /// * `null`
+  /// * [bool]s
+  /// * [num]s
+  /// * [String]s
+  /// * [List]s of supported values
+  /// * [Map]s from strings to supported values
+  static const MessageCodec json = const _JSONCodec();
+
+  /// [MessageCodec] using the Flutter standard binary message encoding.
+  ///
+  /// The standard codec is guaranteed to be compatible with the corresponding
+  /// standard codec for FlutterChannels on the host platform. These two parts
+  /// of the Flutter SDK are evolved synchronously.
+  ///
+  /// Supported messages are acyclic values of these forms:
+  ///
+  /// * `null`
+  /// * [bool]s
+  /// * [num]s
+  /// * [String]s
+  /// * [Uint8List]s, [Int32List]s, [Int64List]s, [Float64List]s
+  /// * [List]s of supported values
+  /// * [Map]s from supported values to supported values
+  static const MessageCodec standard = const _StandardCodec();
+
+  /// Encodes the specified [message] in binary.
+  ///
+  /// Returns `null` if the message is `null`.
+  ByteData encodeMessage(dynamic message);
+
+  /// Decodes the specified [message] from binary.
+  ///
+  /// Returns `null` if the message is `null`.
+  dynamic decodeMessage(ByteData message);
+
+  /// Encodes the specified method call in binary.
+  ///
+  /// The [name] of the method must be non-null. The [arguments] may be `null`.
+  ByteData encodeMethodCall(String name, dynamic arguments);
```

Comment:
As in the Java code review, I think we should disentangle the methodcall-based and message-based codecs.  If necessary, we can have separate `PlatformChannel` classes to drive them.  For example, we could have a `PlatformMessageChannel` and a `PlatformMethodChannel` (or some other better naming scheme.